### PR TITLE
Feature zeroed leaves support

### PR DESCRIPTION
**Pull request by KiselevMaxim**

### CODE REVIEW

**Review by Shchepetov — CHANGES_REQUESTED** [review]

File: contracts/LiquidityTree.sol [hunk 1]
```diff
@@ -516,6 +529,91 @@ contract LiquidityTree {
...
+     * @param node - start from node
+     * @param begin - node left element
+     * @param end - node right element
+     * @param l - left leaf child
+     * @param r - right leaf child
+     * @param amount - amount to add/reduce stored amounts
+     * @param isSub - true means negative to reduce
+     * @return isUpdateInsufficient - true - if can't increase/reduce value because of insufficient or zero value.
+     */
+    function pushLazyPreview(
```
[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol:544]: yes

File: test/liquidityTree-test.js [hunk 2]
```diff
@@ -193,9 +207,9 @@ describe("LiquidityTree", () => {
 
       */
 
-      expect((await sTree.treeNode(4)).amount).to.be.equal("457142857142800000000");
-      expect((await sTree.treeNode(10)).amount).to.be.equal("228571428571438095238");
-      expect((await sTree.treeNode(22)).amount).to.be.equal("114285714285761904762");
+      expect((await sTree.treeNode(4)).amount).to.be.closeTo("457142857142800000000", 100000000);
```
[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply test/liquidityTree-test.js:210]: actually increased, but agree test must be refactored

[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply test/liquidityTree-test.js]: refactored

File: contracts/LiquidityTree.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (Shchepetov) [reply contracts/LiquidityTree.sol:544]: please make the name more precise

File: contracts/LiquidityTree.sol [hunk 3]
```diff
@@ -516,6 +529,91 @@ contract LiquidityTree {
...
+                    pushLazyPreview(
+                        node * 2 + 1,
+                        mid + 1,
+                        end,
+                        mid + 1,
+                        r,
+                        amount - forLeftAmount,
+                        isSub
+                    )
+                ) return true;
```
Comment (Shchepetov) [inline contracts/LiquidityTree.sol:607]: `return pushLazyPreview( node * 2, begin, mid, l, mid, forLeftAmount, isSub) || pushLazyPreview(node * 2 + 1, mid + 1,  end, mid + 1,  r,  amount - forLeftAmount, isSub);`

File: contracts/LiquidityTree.sol [hunk 4]
```diff
@@ -516,6 +529,91 @@ contract LiquidityTree {
...
+            }
+        } else {
+            // [l,r] in [mid+1,end] - all leafs in right child
+            if (
+                pushLazyPreview(node * 2 + 1, mid + 1, end, l, r, amount, isSub)
+            ) return true;
```
Comment (Shchepetov) [inline contracts/LiquidityTree.sol:613]: `return pushLazyPreview(node * 2 + 1, mid + 1, end, l, r, amount, isSub);`

**Review by Shchepetov — CHANGES_REQUESTED** [review]

File: contracts/LiquidityTree.sol [hunk 5]
```diff
@@ -516,6 +529,91 @@ contract LiquidityTree {
...
+        // if node leafs equal to leaf interval then stop
+        if ((begin == l && end == r) || (begin == end)) return false;
+
+        uint48 mid = (begin + end) / 2;
+
+        if (begin <= l && l <= mid) {
+            if (begin <= r && r <= mid) {
+                // [l,r] in [begin,mid] - all leafs in left child
+                if (pushLazyPreview(node * 2, begin, mid, l, r, amount, isSub))
+                    return true;
```
[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol:571]: return must be done only for `true` result, means only if interrupt condition found

File: contracts/LiquidityTree.sol (hunk 3 above)
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol:607]: return must be done only for `true` result, means only if interrupt condition found

Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol:607]: return must be done only for `true` result, means only if interrupt condition found

File: contracts/LiquidityTree.sol (hunk 4 above)
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol:613]: return must be done only for `true` result, means only if interrupt condition found

File: contracts/LiquidityTree.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol]: renamed

File: contracts/LiquidityTree.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol]: fixed

File: contracts/LiquidityTree.sol (hunk 3 above)
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol]: fixed

File: contracts/LiquidityTree.sol (hunk 4 above)
Comment (KiselevMaxim) [reply contracts/LiquidityTree.sol]: fixed

**Review by Shchepetov — APPROVED** [review]

**Pull request merged**